### PR TITLE
Consistently add undefined/missing to optional tuple element types

### LINKED ISSUE #50753: Removing null and undefined from a Tuple breaks equality
# Bug Report
When trying to remove nils from a Tuple it resolves to an Tuple that do not extends to the identical Tuple (as represented by the compiler)

So in the example below the reult of a type is `[a: 1, b?:2]` which do not extends the type `[a: 1, b?:2]` ??

### 🔎 Search Terms
Identical type - Invalid Equality
tuple optional extends

### 🕗 Version & Regression Information
Discovered the issue today and it's referenced on the following [stackoverflow](https://stackoverflow.com/questions/73703051/identical-tuples-are-not-equals)
I did test it in the playground upto version  4.05 of typescript (syntax breaks in version prior)

### ⏯ Playground Link

[playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgdgOQK4BsUEsBmIA8AVKCAD2AjgBMBnKAJwgENyB7OFEKJOAaziYHc4AbQC6APigBeKAG8AsACgoUQQGko6OFC4QQTTFDzCAXFACiRAMYok5CPlXCANFDioUUAD4cKETBojkogoAvgoKoJBQAGJMTJKwiG5YuIL0JgCMzgBGAPwmAEye3rZ+cAFiCgD0lUpQAHo5YfIR0HgQlMDxqRnZeVD5woQkZFTRsVA5UMA0SNAmmPQolBBVNUoNCkA)

### 💻 Code
```ts
type UnNullify<T extends readonly unknown[]> = {
  [K in keyof T]: Exclude<T[K], null | undefined>
}

type Foo = UnNullify<[a: 1, b?: 2 | undefined]>
//   [a: 1, b?: 2]

type Test = [a: 1, b?: 2] extends Foo ? true : false
//   false ??
```

### 🙁 Actual behavior
Two types represented identical by the compiler do not extends each other

### 🙂 Expected behavior
Expected the two types to extends

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #50753.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21637,7 +21637,8 @@ namespace ts {
 
         function getOptionalType(type: Type, isProperty = false): Type {
             Debug.assert(strictNullChecks);
-            return type.flags & TypeFlags.Undefined ? type : getUnionType([type, isProperty ? missingType : undefinedType]);
+            const missingOrUndefined = isProperty ? missingType : undefinedType;
+            return type.flags & TypeFlags.Undefined || type.flags & TypeFlags.Union && (type as UnionType).types[0] === missingOrUndefined ? type : getUnionType([type, missingOrUndefined]);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:21641]: The `types[0]` check always seems extremely fragile, but I suppose it's convenient.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts:21641]: Yeah, but we make that assumption in several other places, so seems fine to do it here as well.

**Comment by ahejlsberg** [comment]
Only one change in the test suites: [reduxjs/reselect](https://github.com/reduxjs/reselect) apparently depends on the ability for mapped types to strip `undefined` from the types of optional properties _without_ stripping optionality at the same time. Since such types are [demonstrably incorrect](https://github.com/microsoft/TypeScript/issues/50753#issuecomment-1251431487) I think this break is acceptable.

**Comment by DanielRosenwasser** [comment]
Clearly this is more correct, and we have no examples of breakages in our test suite, across DefinitelyTyped, internal user test, external user test, or the other top 100; but I can't really track what reselect is doing and where inference is falling back to `{}` in those tests.

@markerikson @lukeapage, if reselect can be modified to accommodate the existing and new behavior, then that would be ideal. Otherwise, if you can provide a minimal repro, we'd welcome a follow-up issue.

**Comment by markerikson** [comment]
Uh. To be honest I really don't have bandwidth to make more changes to the Reselect types any time soon. 

What we've got there now works, and it took me weeks of effort to get it working correctly because of how complicated the mapped function param type extraction is:

https://github.com/reduxjs/reselect/blob/v4.1.5/src/types.ts#L100-L148

I'm going to be pretty unhappy if this gets broken randomly :(

**Comment by ahejlsberg** [comment]
If I understood what `MergeParameters` is doing, I'd be happy to think about an implementation that doesn't break.

**Comment by markerikson** [comment]
Okay, lemme see if I can give a really fast example.

Reselect accepts multiple "input selectors", and an "output selector":

```ts
const selectItemsByCategory = createSelector(
  (state: RootState) => state.items,
  (state: RootState, category: string) => category,
  (items, category => {
    return items.filter(item => item.category === category);
  }
)
```

At runtime, when you call `const items = selectItemsByCategory(state, "food")`, it maps over each input selector and calls it, like:

```ts
const extractedValues = inputSelectors.map(input => input(...args));
return outputSelector(...extractedValues);
```

The goal here is:

- Determine the type of the parameters for each input selector
- Group them by index ("all params at index 0", "all params at index 1")
- For each group, intersect them together so that we have the tightest possible type that is valid at that param index
- Generate a new selector that uses those types as its arguments

To do that, `MergeParamaters`:

- Gets "the type of each input selector"
- Gets "the type of each input selector's params"
- Shuffles them to group by param index
- For each index, runs a recursive intersection across the params at that index
- Turns all that back into `[typeForIndex0, typeForIndex1, typeForIndex2]`, etc

**Comment by DanielRosenwasser** [comment]
I think the best step for us to try to bring this change in for the 4.9 beta - that way people can at least try it out easily. If there is no way to resolve the current issues with reselect, we can discuss further options.

**Comment by ahejlsberg** [comment]
@markerikson Thanks for the explanation. Below is a simplified version of `MergeParameters` that works with TS version 4.7 and later. Hope this helps.

```ts
type UnknownFunction = (...args: any[]) => any;

type LongestTuple<T extends readonly unknown[][]> =
    T extends [infer U extends unknown[]] ? U :
    T extends [infer U, ...infer R extends unknown[][]] ? MostProperties<U, LongestTuple<R>> :
    never;

type MostProperties<T, U> = keyof U extends keyof T ? T : U;

type ElementAt<T extends unknown[], N extends keyof any> = N extends keyof T ? T[N] : unknown;

type ElementsAt<T extends readonly unknown[][], N extends keyof any> = {
    [K in keyof T]: ElementAt<T[K], N>
}

type Intersect<T extends readonly unknown[]> =
    T extends [] ? unknown :
    T extends [infer H, ...infer T] ? H & Intersect<T> :
    T[number];

type MergeTuples<T extends readonly unknown[][], L extends unknown[] = LongestTuple<T>> = {
    [K in keyof L]: Intersect<ElementsAt<T, K>>
}

type ExtractParameters<T extends readonly UnknownFunction[]> = {
    [K in keyof T]: Parameters<T[K]>
}

export type MergeParameters<T extends readonly UnknownFunction[]> = "0" extends keyof T ? MergeTuples<ExtractParameters<T>> : Parameters<T[number]>;

type Test1 = MergeParameters<[(state: unknown, a: string) => void, (state: unknown) => void, (state: unknown, a: "a" | "b", b?: number) => void]>;
type Test2 = MergeParameters<((state: unknown, a: string) => void)[]>;
```

**Comment by markerikson** [comment]
Following up on this:

@ahejlsberg , yes, that implementation does appear to work right _ **_thank_ you!**

The bad news is that now I have to figure out how to ship this in Reselect and still support users on TS 4.6 and earlier, and _that_ is a separate pain point that I just discussed over in https://github.com/microsoft/TypeScript/issues/50196#issuecomment-1301609232 .  (There's a hacky workaround that involves build-time shenanigans that technically works, but is _not_ a satisfying solution.)

**Comment by ssalbdivad** [comment]
I ran into this problem while working on a utility function for combining trait-like functions and implemented a similar type before someone mentioned this one existed.

This is just the simplified form for a binary intersection, but could easily be composed together to form a "merge" generic like @ahejlsberg's. There are so many edge cases when it comes to this but it seems to handle most well, e.g. a trailing rest parameter or merging optional and required parameters in the same position.

Given how tricky this is to implement at a type-level, would be awesome if TS could compute this directly when intersecting tuples or provide some way to access it. There'd be some nuance since the rules would work differently depending on whether the tuples represented parameters or not, but as is it seems extremely hard (maybe impossible?) to get everything right using types alone, especially when you consider stuff like `(state: unknown, ...rest: [...string[], number]) => void`.

```ts
type intersectParameters<
	l extends readonly unknown[],
	r extends readonly unknown[],
	result extends readonly unknown[] = []
> = l extends readonly []
	? [...result, ...r]
	: r extends readonly []
	? [...result, ...l]
	: [number, number] extends [l["length"], r["length"]]
	? [...result, ...(l[number] & r[number])[]]
	: [l, r] extends [
			readonly [unknown?, ...infer lTail],
			readonly [unknown?, ...infer rTail]
	  ]
	? intersectParameters<lTail, rTail, [...result, l[0] & r[0]]>
	: l extends readonly [infer lHead, ...infer lTail]
	? intersectParameters<lTail, r, [...result, lHead & r[number]]>
	: r extends readonly [infer rHead, ...infer rTail]
	? intersectParameters<l, rTail, [...result, l[number] & rHead]>
	: result
```

**Comment by ssalbdivad** [comment]
Welp, I found a few more edge cases. Here's the updated version if you want your parameter intersections to be more precise in dealing with optionals (I added some comments this time because things got a little out of hand trying to cover all the edge cases 😆):

```ts
type parseNextParam<params extends readonly unknown[]> =
	params extends readonly []
		? {
				// Since a longer args array is assignable to a shorter one,
				// elements after the last are typed as unknown. For a normal
				// tuple (not parameters), this would be never.
				head: unknown
				optional: false
				tail: []
				done: true
		  }
		: params extends readonly [(infer head)?, ...infer tail]
		? [tail, params] extends [params, tail]
			? {
					head: head
					optional: true
					tail: tail
					done: true
			  }
			: {
					head: head
					optional: params[0] extends head ? false : true
					tail: tail
					done: false
			  }
		: never

type ParameterParseResult = {
	head: unknown
	optional: boolean
	tail: readonly unknown[]
	done: boolean
}

type intersectParameters<
	l extends readonly unknown[],
	r extends readonly unknown[],
	result extends readonly unknown[] = []
> = [parseNextParam<l>, parseNextParam<r>] extends [
	infer lState extends ParameterParseResult,
	infer rState extends ParameterParseResult
]
	? false extends lState["done"] | rState["done"]
		? // while either array still has fixed elements remaining, continue intersecting heads
		  intersectParameters<
				lState["tail"],
				rState["tail"],
				[
					...result,
					// the intersection is optional only if both elements are
					...(lState["optional"] | rState["optional"] extends true
						? [(lState["head"] & rState["head"])?]
						: [lState["head"] & rState["head"]])
				]
		  >
		: // once both arrays have reached their fixed end or a variadic element, return the final result
		  [
				...result,
				// if the state is done and non-optional, we've reached the end of a non-variadic tuple
				...(lState["optional"] extends false
					? rState["optional"] extends false
						? []
						: // if done and non-optional, we've reached a variadic element
						  // (or it's just a normal array, since number[] === [...number[]])
						  rState["tail"]
					: rState["optional"] extends false
					? lState["tail"]
					: // if we've reached a variadic element in both arrays, intersect them
					  (lState["head"] & rState["head"])[])
		  ]
	: never
```

**Pull request merged**